tests/TestHarness/launcher.py

```diff
@@ -66,9 +188,264 @@ def comma_separated(string):
+        for node_name, node in self.network.nodes.items():
+            is_bios = node_name == 'bios'
+            node.keys.append(KeyStrings('EOS6MRyAjQq8ud7hVNYcfnVPJqcVpscN5So8BhtHuGYqET5GDW5CV',
+                                         '5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3'))
+            if is_bios:
+                node.producers.append('eosio')
+            else:
+                if i < non_bios:
+                    count = per_node
+                    if extra:
+                        count += 1
+                        extra -= 1
+                    while count:
+                        prodname = producer_name(producer_number)
+                        node.producers.append(prodname)
+                        producer_number += 1
+                        count -= 1
+                    for j in range(0, self.args.shared_producers):
+                        prodname = producer_name(j, True)
+                        node.producers.append(prodname)
+                node.dont_start = i >= to_not_start_node
+            if not is_bios:
+                i += 1
+
+
+    def generate(self):
+        {
+            'ring': self.make_line,
+            'line': lambda: self.make_line(False),
+            'star': self.make_star,
+            'mesh': self.make_mesh,
+        }.get(self.args.shape, self.make_custom)()
+
+        if not self.args.nogen:
+            for node_name, node in self.network.nodes.items():
+                node.config_dir_name.mkdir(parents=True)
+                self.write_config_file(node)
+                self.write_logging_config_file(node)
+                self.write_genesis_file(node)
+        
+        self.write_dot_file()
+
+    def write_config_file(self, node):
+        is_bios = node.name == 'bios'
+        peers = '\n'.join([f'p2p-peer-address = {self.network.nodes[p].p2p_endpoint}' for p in node.peers])
+        if len(node.producers) > 0:
+            producer_keys = 'private-key = ["EOS6MRyAjQq8ud7hVNYcfnVPJqcVpscN5So8BhtHuGYqET5GDW5CV","5KQwrPbwdL6PhXujxW37FSSQZ1JiwsST4cqQzDeyXtP79zkvFD3"]\n'
+            producer_names = '\n'.join([f'producer-name = {p}' for p in node.producers])
+            producer_plugin = 'plugin = eosio::producer_plugin\n'
+        else:
+            producer_keys = ''
+            producer_names = ''
+            producer_plugin = ''
+        config = f'''blocks-dir = {block_dir}
+http-server-address = {node.host_name}:{node.http_port}
+http-validate-host = false
+p2p-listen-endpoint = {node.listen_addr}:{node.p2p_port}
+p2p-server-address = {node.p2p_endpoint}
+{"enable-stale-production = true" if is_bios else ""}
+{"p2p-peer-address = %s" % self.network.nodes['bios'].p2p_endpoint if not is_bios else ""}
+{peers}
+{producer_keys}
+{producer_names}
+{producer_plugin}
+plugin = eosio::net_plugin
+plugin = eosio::chain_api_plugin
+'''
+        with open(node.config_dir_name / 'config.ini', 'w') as cfg:
```

Comment:
Should the file open come before all the processing to create the config str? If the open fails, that processing is for naught.